tests/unit/s2n_server_new_session_ticket_test.c

```diff
@@ -303,4 +304,181 @@ int main(int argc, char **argv)
+
+            s2n_blocked_status blocked = 0;
+            EXPECT_OK(s2n_tls13_server_nst_send(conn, &blocked));
+
+            /* Check five records were written */
+            uint16_t record_len = 0;
+            for (size_t i = 0; i < tickets_to_send; i++) {
+                EXPECT_SUCCESS(s2n_stuffer_skip_read(&stuffer, RECORD_LEN_MARKER));
+                EXPECT_SUCCESS(s2n_stuffer_read_uint16(&stuffer, &record_len));
+                EXPECT_TRUE(record_len > 0);
+                EXPECT_SUCCESS(s2n_stuffer_skip_read(&stuffer, record_len));
+            }
+            EXPECT_TRUE(s2n_stuffer_data_available(&stuffer) == 0);
+
+            /* No more tickets to send */
+            EXPECT_SUCCESS(s2n_stuffer_rewrite(&stuffer));
+            EXPECT_OK(s2n_tls13_server_nst_send(conn, &blocked));
+            EXPECT_EQUAL(0, s2n_stuffer_data_available(&stuffer));
+
+            EXPECT_SUCCESS(s2n_stuffer_free(&stuffer));
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_config_free(config));
+        }
+    }
+
+    /* Functional test: s2n_negotiate sends new session tickets after the handshake is complete */
+    {
+        /* Setup connections */
+        struct s2n_connection *client_conn, *server_conn;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        /* Setup config */
+        struct s2n_cert_chain_and_key *chain_and_key;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+                S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN, S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+        struct s2n_config *config;
+        EXPECT_NOT_NULL(config = s2n_config_new());
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+        EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(config, 1));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        EXPECT_SUCCESS(s2n_setup_test_keys(server_conn, config));
+
+        uint16_t tickets_to_send = 5;
+        server_conn->tickets_to_send = tickets_to_send;
+
+        struct s2n_stuffer client_to_server;
+        struct s2n_stuffer server_to_client;
+
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&client_to_server, 0));
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&server_to_client, 0));
+
+        EXPECT_SUCCESS(s2n_connection_set_io_stuffers(&server_to_client, &client_to_server, client_conn));
+        EXPECT_SUCCESS(s2n_connection_set_io_stuffers(&client_to_server, &server_to_client, server_conn));
+
+        /* Do handshake */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        /* Check five records were written */
```

Comment:
I think this comment is misleading. More than five records were written during negotiation, the five NewSessionTickets were just the only ones not read as part of the handshake. This might not be true / might need to be updated if we start reading NewSessionTickets at the end of s2n_negotiate too.

Reply:
I edited the comment but yes, this behavior might change in the future.